### PR TITLE
Ability to provide a custom factory

### PR DESCRIPTION
**Pull request by martinmaillard**
Addresses #26 

I tried implementing this because I need it in rdf-validate-shacl.

I'm really not sure about the implementation *or* the tests, so I would love some feedback.

~~One thing that's still missing is that terms coming from the `namespaces` module don't use the provided factory. I'm going to have to fix that.~~

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 const Clownface = require('./lib/Clownface')
 
-function factory ({ dataset, graph, term, value, _context }) {
-  return new Clownface({ dataset, graph, term, value, _context })
+function factory ({ dataset, graph, term, value, factory, _context }) {
+  return new Clownface({ dataset, graph, term, value, factory, _context })
```
Comment (tpluscode) [inline index.js:4]: Doesn't the `factory` variable conflict with the function name?

File: lib/term.js [hunk 2]
```diff
@@ -30,23 +30,23 @@ function literal (value, languageOrDatatype) {
   return term
 }
 
-function namedNode (value) {
+function namedNode (value, factory = rdf) {
```
Comment (tpluscode) [inline lib/term.js:33]: I don't think it makes sense to place a parameter on every method. You would always use `this.factory` IMO

Comment (martinmaillard) [reply lib/term.js:33]: You mean not have a default value?

File: index.js (hunk 1 above)
Comment (martinmaillard) [reply index.js:4]: I don't think that's an issue

File: lib/term.js (hunk 2 above)
Comment (tpluscode) [reply lib/term.js:33]: I mean no parameter at all

```diff
-function namedNode (value, factory = rdf) {
+function namedNode (value) {

-return factory.namedNode(value)
+return this.factory.namedNode(value)
```

**Comment by tpluscode** [comment]
Regarding the overall shape, the design is that to be able can create one instance from another

```js
const graph = clownface({ dataset, term, factory })

const another = clownface(graph)
```

A test would be good to ensure that the factory is retained from the original object

**Comment by martinmaillard** [comment]
That was a good remark. I added a test and noticed it wasn't actually working as intended ;)

File: lib/term.js (hunk 2 above)
Comment (martinmaillard) [reply lib/term.js:33]: But these functions are standalone. They don't have access to the context (`this`).

Comment (tpluscode) [reply lib/term.js:33]: aha, sorry, I missed your reply. thought it was on the Clownface object

then I agree with what you said before. the default seems redundant

Comment (bergos) [reply lib/term.js:33]: Yes, it's a standalone, but the factory argument should be always provided, so there should be no default value for it.

Comment (bergos) [reply lib/term.js:33]: That should be the case for all functions in `lib/term.js`.

**Comment by martinmaillard** [comment]
Ok will change that.

And what should I do about the namespaces? Should I add a `.ns` on the Clownface/Context objects?

File: lib/term.js (hunk 2 above)
Comment (martinmaillard) [reply lib/term.js:33]: That means `term` has to be called like this in some cases: `term('test', undefined, undefined, factory)`. Should the factory be the first param? Or should we abandon the idea of having pure functions for this module and move them to a class?

**Comment by bergos** [comment]
The namespaces defined in [lib/namespace.js](https://github.com/rdf-ext/clownface/blob/master/lib/namespace.js)? They are also used to create quads in the `addList` method, so the given factory should be used. Maybe wrap the `namespace.js` code into a function that returns the namespace and assign the namespace to the `Context` object. Maybe also worth to change the `Context` constructor to use named arguments. [This repeating pattern](https://github.com/rdf-ext/clownface/blob/master/lib/Context.js#L16) could be changed to use a `clone` function like this:

```javascript
clone ({ dataset = this.dataset, graph = this.graph, term = this.term, namespace = this.namespace } = {}) {
  return new Context({ dataset, graph, term, namespace })
}
```

**Comment by martinmaillard** [comment]
I made the `namespace` changes. There's still the `term` stuff to figure out.

**Comment by bergos** [comment]
> There's still the term stuff to figure out.

Which topic exactly? Anything required from my side?

**Comment by martinmaillard** [comment]
It's related to [this comment](https://github.com/rdf-ext/clownface/pull/31#discussion_r405624394). Would love your feedback!

File: lib/term.js (hunk 2 above)
Comment (bergos) [reply lib/term.js]: Let's keep them as pure functions.

**Comment by martinmaillard** [comment]
Alright, then on my side it's ready.

**Pull request merged**